pkg/driver/aws-ebs/aws_ebs_tags_controller.go

```diff
@@ -0,0 +1,441 @@
+	}
+	return f.Name(), nil
+}
+
+// getInfrastructure retrieves the Infrastructure resource in OpenShift
+func (c *EBSVolumeTagsController) getInfrastructure(ctx context.Context) (*configv1.Infrastructure, error) {
+	backoff := wait.Backoff{
+		Duration: operationDelay,
+		Factor:   operationBackoffFactor,
+		Steps:    operationRetryCount,
+	}
+	infra := &configv1.Infrastructure{}
+	err := wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		var apiError error
+		infra, apiError = c.commonClient.ConfigInformers.Config().V1().Infrastructures().Lister().Get(infrastructureName)
+		if apiError != nil {
+			klog.Errorf("error listing infrastructures objects: %v", apiError)
+			return false, nil
+		}
+		if infra != nil {
+			return true, nil
+		}
+		return false, nil
+	})
+	return infra, err
+}
+
+func (c *EBSVolumeTagsController) getEBSCloudCredSecret(ctx context.Context) (*v1.Secret, error) {
+	backoff := wait.Backoff{
+		Duration: operationDelay,
+		Factor:   operationBackoffFactor,
+		Steps:    operationRetryCount,
+	}
+	var awsCreds *v1.Secret
+	err := wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		var apiError error
+		awsCreds, apiError = c.commonClient.KubeClient.CoreV1().Secrets(awsEBSSecretNamespace).Get(ctx, awsEBSSecretName, metav1.GetOptions{})
+		if apiError != nil {
+			klog.Errorf("error getting secret object: %v", apiError)
+			return false, nil
+		}
+		if awsCreds != nil {
+			return true, nil
+		}
+		return false, nil
+	})
+	return awsCreds, err
+
+}
+
+// processInfrastructure processes the Infrastructure resource and updates EBS tags
+func (c *EBSVolumeTagsController) processInfrastructure(ctx context.Context, infra *configv1.Infrastructure, ec2Client *ec2.EC2) error {
+	if infra.Status.PlatformStatus != nil && infra.Status.PlatformStatus.AWS != nil &&
+		infra.Status.PlatformStatus.AWS.ResourceTags != nil {
+		awsInfra := infra.Status.PlatformStatus.AWS
+		err := c.fetchPVsAndUpdateTags(ctx, awsInfra.ResourceTags, ec2Client)
+		if err != nil {
+			klog.Errorf("Error processing PVs for infrastructure update: %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+// fetchPVsAndUpdateTags retrieves all PVs and updates the AWS EBS tags in batches of 100
+func (c *EBSVolumeTagsController) fetchPVsAndUpdateTags(ctx context.Context, resourceTags []configv1.AWSResourceTag, ec2Client *ec2.EC2) error {
+	pvs, err := c.listPersistentVolumesWithRetry(ctx)
+	if err != nil {
+		return fmt.Errorf("error fetching PVs: %v", err)
+	}
+	// Compute the hash for the new set of tags
+	newTagsHash := computeTagsHash(resourceTags)
+	pvsToBeUpdated := filterUpdatableVolumes(pvs, newTagsHash)
+
+	// If there are no volumes to update, return early
+	if len(pvsToBeUpdated) == 0 {
+		klog.Infof("No volume tags to update as hashes are unchanged")
+		return nil
+	}
+
+	// Process the volumes in batches
```

Comment:
I mentioned in my original comment that, the main controller should not try and update PVs which will be processed by failed worker queue. I do not see that scenairo handled here.